### DISCUSSION TITLE
Flux restarting pods again after rolling restart

### DISCUSSION THREAD

**Discussion by jwerre** [fluxcd/flux2#2755] [Q&A] [answered]
I'm running into a strange issue where if I do a rolling restart on a deployment the kustomize-controller dispatches an event (after 10 or so minutes ) and restarts that same deployment. For example:

```
kubectl rollout restart deployment -n MY_NS MY_DEP
```

```
flux logs --since 30m --kind Kustomization

[...]

2022-05-20T12:54:57.993Z info Kustomization/flux-system.flux-system - server-side apply completed 
2022-05-20T12:54:59.038Z info Kustomization/flux-system.flux-system - server-side apply completed
2022-05-20T12:54:59.058Z info Kustomization/flux-system.flux-system - Reconciliation finished in 2.033319648s, next run in 10m0s

2022-05-20T13:05:00.044Z info Kustomization/flux-system.flux-system - server-side apply completed
2022-05-20T13:05:01.152Z info Kustomization/flux-system.flux-system - server-side apply completed
2022-05-20T13:05:01.156Z info Kustomization/flux-system.flux-system - Dispatching event: Deployment/MY_NS/MY_DEP configured
2022-05-20T13:05:01.177Z info Kustomization/flux-system.flux-system - Reconciliation finished in 2.114664112s, next run in 10m0s
2022-05-20T13:05:01.178Z info Kustomization/flux-system.flux-system - Dispatching event: Reconciliation finished in 2.114664112s, next run in 10m0s

[... MY_DEP restarts here]

2022-05-20T13:15:02.471Z info Kustomization/flux-system.flux-system - server-side apply completed
2022-05-20T13:15:04.037Z info Kustomization/flux-system.flux-system - server-side apply completed
2022-05-20T13:15:04.090Z info Kustomization/flux-system.flux-system - Reconciliation finished in 2.910387883s, next run in 10m0s
```

After Kustomization reconciliation finishes at 13:05:01 `MY_DEP` preforms another rolling restart. Any idea why this could be happening or how to fix it? I've restarted all my flux deployments and rerun `flux bootstrap` which doesn't seem to resolve the issue. Below is all the information on my cluster:

```
flux check
► checking prerequisites
✔ Kubernetes 1.21.9-eks-0d102a7 >=1.20.6-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.21.0
✔ image-automation-controller: deployment ready
► ghcr.io/fluxcd/image-automation-controller:v0.22.1
✔ image-reflector-controller: deployment ready
► ghcr.io/fluxcd/image-reflector-controller:v0.18.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.25.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.23.5
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.24.4
✔ all checks passed
```

```
flux --version
flux version 0.30.2
```

```
kubectl describe -n flux-system deployments.apps kustomize-controller
Name:                   kustomize-controller
Namespace:              flux-system
CreationTimestamp:      Thu, 27 Jan 2022 09:29:58 -0700
Labels:                 app.kubernetes.io/instance=flux-system
                        app.kubernetes.io/part-of=flux
                        app.kubernetes.io/version=v0.30.2
                        control-plane=controller
                        kustomize.toolkit.fluxcd.io/name=flux-system
                        kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:            deployment.kubernetes.io/revision: 6
Selector:               app=kustomize-controller
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           app=kustomize-controller
  Annotations:      prometheus.io/port: 8080
                    prometheus.io/scrape: true
  Service Account:  kustomize-controller
  Containers:
   manager:
    Image:       ghcr.io/fluxcd/kustomize-controller:v0.25.0
    Ports:       8080/TCP, 9440/TCP
    Host Ports:  0/TCP, 0/TCP
    Args:
      --events-addr=http://notification-controller.flux-system.svc.cluster.local./
      --watch-all-namespaces=true
      --log-level=info
      --log-encoding=json
      --enable-leader-election
    Limits:
      cpu:     1
      memory:  1Gi
    Requests:
      cpu:      100m
      memory:   64Mi
    Liveness:   http-get http://:healthz/healthz delay=0s timeout=1s period=10s #success=1 #failure=3
    Readiness:  http-get http://:healthz/readyz delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:
      RUNTIME_NAMESPACE:   (v1:metadata.namespace)
    Mounts:
      /tmp from temp (rw)
  Volumes:
   temp:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  <none>
NewReplicaSet:   kustomize-controller-784bd54978 (1/1 replicas created)
Events:          <none>
```

---------------

EDIT:

Perhaps this is a different issue but I noticed there is one error in my logs regarding the `kustomize-controller`:

```
2022-05-14T15:54:57.203Z error Kustomization/flux-system.flux-system - Reconciliation failed after 32.983979ms, next try in 10m0s failed to download artifact from http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/SOME_HASH.tar.gz, status: 404 Not Found
```

This is about the time all this started and I wonder if it has something to do with it. So I saw [issue #2307](https://github.com/fluxcd/flux2/discussions/2307) and thought it could be a pod to pod communication issue (100% not an overload issue) so  I create and logged into a pod in the `flux-system` namespace and did a curl on this url:

```
$ curl -vvv http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/SOME_HASH.tar.gz

*   Trying 172.20.153.209:80...
* Connected to source-controller.flux-system.svc.cluster.local (172.20.153.209) port 80 (#0)
> GET /gitrepository/flux-system/flux-system/SOME_HASH.tar.gz HTTP/1.1
> Host: source-controller.flux-system.svc.cluster.local
> User-Agent: curl/7.79.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Content-Type: text/plain; charset=utf-8
< X-Content-Type-Options: nosniff
< Date: Thu, 19 May 2022 22:51:14 GMT
< Content-Length: 19
<
404 page not found
* Connection #0 to host source-controller.flux-system.svc.cluster.local left intact
```

It looks like it connect successfully to the host:
```
Connected to source-controller.flux-system.svc.cluster.local (172.20.153.209) port 80 (#0)"
```

... but failed to find `SOME_HASH.tar.gz` and returned a page not found error.

EDIT:

**Comment by stefanprodan** [comment, marked as answer]
Flux does a rollback of any kubectl edits in-cluster, `kubectl rollout restart deployment` adds an annotation that Flux removes, so Kubernetes sees the removal and restarts the deployment a 2nd time. All manual edits must be using `kubectl --field-manager=flux-client-side-apply ` docs here: https://fluxcd.io/docs/faq/#why-are-kubectl-edits-rolled-back-by-flux

**Reply by jwerre** [reply]
I don't know how I haven't noticed that before... is this new?

**Reply by stefanprodan** [reply]
Not that new, the kubectl rollback feature was released in late March.

**Reply by beasteers** [reply]
Does it have to be literally `--field-manager=flux-client-side-apply` or can it be any custom manager value (e.g. the name of an operator)? 

I'm submitting a bug report to an operator about this because flux keeps removing their Lock annotations, but the idea of having them implement Flux-specific functionality doesn't seem like a simple ask.

**Reply by stefanprodan** [reply]
Yes it can be some other controller name, but not kubectl.

**Reply by beasteers** [reply]
In the meantime, does flux have any resource-side ways of controlling this? 

For example, I just found ArgoCD docs describing `preserveFields` letting you whitelist annotations.

https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Controlling-Resource-Modification/#preserving-changes-made-to-an-applications-annotations-and-labels

**Reply by matheuscscp** [reply]
Maybe you're looking for this: https://fluxcd.io/flux/components/kustomize/kustomizations/#controlling-the-apply-behavior-of-resources

**Reply by BradErz** [reply]
@stefanprodan Checking the code and implementation of `kubectl rollout restart deployment` it is adding a [field-manager](https://kubernetes.io/docs/reference/kubectl/generated/kubectl_rollout/kubectl_rollout_restart/#options:~:text=jsonpath%20output%20formats.-,%2D%2Dfield%2Dmanager,-string%C2%A0%C2%A0%C2%A0%C2%A0%C2%A0Default%3A%20%22kubectl) (from the code it looks to have been there for 5 years.) which by default is set to `kubectl-rollout`. However, during my testing, if I run the rollout command and then directly afterwards run a flux `flux reconcile ks $ks_name`, the deployment is restarted again to remove the annotation and the managed field. It does work if i supply the `--field-manager=flux-client-side-apply`. Maybe if the field manager worked, this could be a valid solution.

<details>

<summary>Related versions</summary>

```bash
kubectl version
Client Version: v1.32.2
Kustomize Version: v5.5.0
Server Version: v1.30.9-gke.1009000
WARNING: version difference between client (1.32) and server (1.30) exceeds the supported minor version skew of +/-1
```

```bash
 flux version
flux: v2.4.0
distribution: flux-v2.3.0
helm-controller: v1.0.1
image-automation-controller: v0.38.0
image-reflector-controller: v0.32.0
kustomize-controller: v1.3.0
notification-controller: v1.3.0
source-controller: v1.3.0
```

</details>


Overall, I think the developer experience of users running `kubectl rollout restart deployment` is heavily impacted by this. Who is supposed to remember to put `--field-manager=flux-client-side-apply` every time? And communicating other users of the cluster most will never remember actually to do that.

Other controllers are also attempting to restart deployments through similar mechanisms. For example, the Vault Secrets Operator adds the annotation [vso.secrets.hashicorp.com/restartedAt](https://github.com/hashicorp/vault-secrets-operator/blob/c9e7d1375dc61c22e60f6d980c10b7747c76cf2b/helpers/rollout_restart.go#L25). 

@matheuscscp, I don't think what you mentioned is the same as the argo `preserveFields`. My understanding of the feature you linked would mean other side effects. For example, the document talks about a more general "edit" behaviour, which we don't want to allow, and those things should be reset by flux itself. Imagine someone adds a new definition to a deployment object in the cluster which doesn't exist in the git definition of that deployment. If we swapped the `ssa` policy to [merge](https://fluxcd.io/flux/components/kustomize/kustomizations/#merge), it wouldn't be removed.

In the Argo preserve fields mechanism, we would be able to specify:
```yaml
  preservedFields:
    annotations:
    - "kubectl.kubernetes.io/restartedAt"
    - "vso.secrets.hashicorp.com/restartedAt"
```

I personally think this mechanism allows controllers to stay independent of each other while allowing operators to easily decide which annotations/labels should be maintained during a reconciliation.

**Reply by BradErz** [reply]
I guess checking more into the code these field managers are automatically ignored.
https://github.com/fluxcd/kustomize-controller/blob/444b8d265e2ecd10f63b2ae8a6bf95072bec2c86/internal/controller/kustomization_controller.go#L708-L717

And this code is checking for a prefix of the manager name:
https://github.com/fluxcd/pkg/blob/00d1ceb1c10ac3bcbb8d5793bc3aba176bfaec40/ssa/patch.go#L127

So the `"manager": "kubectl-rollout"` added by `kubectl rollout restart` is removed. 

I guess that means atleast if other operators such as Vault Secrets Operator added a field-manager when they apply their annotation it would work. But currently its not possible to use `kubectl rollout restart` without specifying a different field manager. Which i hope you agree is kind of annoying.

Im happy to work on a patch for this if you're willing to discuss/accept a solution. Im also happy to go and try to patch the vault secrets operator to add the field-manager if thats the main direction flux wants to take with this.
